src/Analyser/SpecifiedTypes.php

```diff
@@ -43,59 +41,52 @@ public function getSureNotTypes(): array
 		return $this->sureNotTypes;
 	}
 
-	public function addSureType(Node $expr, string $exprString, Type $type): self
+	public function intersectWith(SpecifiedTypes $other): self
 	{
-		$types = $this->sureTypes;
-		if (isset($types[$exprString])) {
-			unset($types[$exprString]); // because we don't have intersection types
-			return new self($types, $this->sureNotTypes);
-		}
-
-		$types[$exprString] = [
-			$expr,
-			$type,
-		];
-
-		return new self(
-			$types,
-			$this->sureNotTypes
-		);
-	}
+		$sureTypeUnion = [];
+		$sureNotTypeUnion = [];
 
-	public function addSureNotType(Node $expr, string $exprString, Type $type): self
-	{
-		$types = $this->sureNotTypes;
-		if (isset($types[$exprString])) {
-			$type = $types[$exprString][1]->combineWith($type);
+		foreach ($this->sureTypes as $exprString => list($exprNode, $type)) {
+			if (isset($other->sureTypes[$exprString])) {
+				$sureTypeUnion[$exprString] = [
+					$exprNode,
+					$type->combineWith($other->sureTypes[$exprString][1]),
+				];
+			}
 		}
 
-		$types[$exprString] = [
-			$expr,
-			$type,
-		];
+//		foreach ($this->sureNotTypes as $exprString => list($exprNode, $type)) {
```

Comment:
I don't want to merge PR with commented out code :( Can you fix it or remove it? Thanks.